### PR TITLE
1211: Speech Training Exercise

### LINKED ISSUE #1211: [T41K-20] Learning mode "Sprachpraxis"
### Describe the Problem

We want to implement the learning mode "Sprachpraxis". 

### Describe Your Preferred Solution

Users see an image and the speak the right word into their smartphone's microphone. The app tells them if it was right or wrong. Users can repeat x times until it is correct.

### PR DESCRIPTION
**Pull request by LeandraH**
### Short Description

This PR implements the Sprachtraining. It uses built-in voice recognition because the models from Vosk and Whisper that I tested were either so big that the apps kept crashing or so small that they couldn't handle rare words like "der Mörser". Unfortunately, that means that on an Android phone without Google, the function isn't available. I also checked out the package [`expo-speech-recognition`](https://github.com/jamsch/expo-speech-recognition) (which would have been a nice reason to look into using some expo in our app), but it seems to also use Google's voice recognition.

### Proposed Changes

- Added a native module for Android and iOS each that handles the voice recognition
- Added a `SpeechTrainingScreen` using that native module
- Added a `SpeechMatchingService` that fine-tunes the checking for looooots of edge cases
- Lots of tests for those edge cases

### Side Effects

- Android users without Google ~~can't use it~~ might have a worse speech recognition service installed
- On the line between being too generous and too strict, I decided to go with being too generous because we want to encourage our users to keep speaking

### Testing

How to test:

I have not managed to get the iOS simulator to recognize my voice, so iOS seems to only work on a real iPhone, which I tested pretty thoroughly

The Android emulator does recognize my voice, after adding German as a language in the phone settings > system > languages > speech > voice input and on-device recognition, plus in the extended controls of the emulator app (the three dots) > microphone > enable host microphone access. However, it's still a little unreliable, so testing on a real Android device is probably better

If you enable dev mode in the app, the screen shows you what the correct answer would be, which I have found helpful

What to test:
Test the Sprachtraining in all possibilities that you can think of (speak clearly, mumble, don't say anything, say the wrong thing, the wrong article, a whole sentence with the word in it, ...)

### Resolved Issues

Fixes: #1211

---

### CODE REVIEW

File: modules/speech-to-text/android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  buildTypes {
+    release {
+      minifyEnabled false
+    }
+  }
+
+  lint {
+    disable "GradleCompatible"
+  }
```
Comment (david-venhoff) [inline modules/speech-to-text/android/build.gradle]:
```suggestion
```

I think these fields can be removed

**Review by david-venhoff — COMMENTED** [review]
Very cool 🎉 
Awesome that you found a way to make it work with the local apis :)

Tested on android emulator.
 
I added some comments, mostly questions and nits though.

I think commit ff453a4402e6e9780a8e6ea20618ad1e9e9ac191 is not needed. I reverted it and got no build errors. Also my experiment ([experiment/native-api-voice-recognition](https://github.com/digitalfabrik/lunes-app/tree/experiment/native-api-voice-recognition)) worked without it.

It might be nice if we could detect on android whether German is installed ([there is an api for that](https://developer.android.com/reference/android/speech/RecognizerIntent#ACTION_GET_LANGUAGE_DETAILS), or we could handle the error when a language is not available) and then make a more user friendly error message.

Also this does not need google play services on android, just a speech to text provider. The only potential problem is that the quality of other speech to text providers can vary

**Comment by LeandraH** [comment]
> I think commit [ff453a4](https://github.com/digitalfabrik/lunes-app/commit/ff453a4402e6e9780a8e6ea20618ad1e9e9ac191) is not needed. I reverted it and got no build errors. Also my experiment ([experiment/native-api-voice-recognition](https://github.com/digitalfabrik/lunes-app/tree/experiment/native-api-voice-recognition)) worked without it.

Oooh, awesome, you're right, they're not needed!

> It might be nice if we could detect on android whether German is installed ([there is an api for that](https://developer.android.com/reference/android/speech/RecognizerIntent#ACTION_GET_LANGUAGE_DETAILS), or we could handle the error when a language is not available) and then make a more user friendly error message.

That's a great idea! I'll get to that, and also update the error message. Including that we apparently don't need Google Play 🥳 

> Also this does not need google play services on android, just a speech to text provider. The only potential problem is that the quality of other speech to text providers can vary

File: src/routes/training/services/SpeechMatchingService.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  const normalizedWord = normalizeText(word)
+  const normalizedTranscript = normalizeText(transcript)
+
+  if (
+    stringSimilarity(normalizedTranscript, normalizedFull) >= FULL_PHRASE_SIMILARITY_THRESHOLD ||
+    isTruncatedPhrase(normalizedTranscript, normalizedFull) ||
+    isTruncatedJoinedPhrase(normalizedTranscript, normalizedFull) ||
+    isMorphemeBoundaryTruncation(normalizedTranscript, normalizedFull) ||
+    containsPhraseTokens(normalizedTranscript, normalizedFull) ||
+    stringSimilarity(normalizedTranscript, normalizedWord) >= WORD_ONLY_SIMILARITY_THRESHOLD
```
[the remark this answers is not in the document]
Comment (LeandraH) [reply src/routes/training/services/SpeechMatchingService.ts]: Hmm, for `isTruncatedJoinedPhrase`, it would work but I didn't see the case of the article being dropped when the recognizer split up compound words.

`isMorphemeBoundaryTruncation` needs at least two tokens, so calling it without the article is rarely useful.

`containsPhraseTokens` could do something but that case is already caught by `stringSimilarity(normalizedTranscript, normalizedWord).`

**Review by david-venhoff — APPROVED** [review]
Thanks, LGTM!

I am still a bit worried that users will not find the correct screen to install German on android, but that should not be a blocker I think. If it turns out to be an problem, we can just open an issue for that :)
